### PR TITLE
RDM-5709-related-to-5398

### PR DESCRIPTION
**Pull request by Thor-tech-of-metal**
RDM-5709: Previously uploaded Document links are lost when uploading additional documents in an event
    [RDM-5709] (https://tools.hmcts.net/jira/browse/RDM-5709).
    related issue:
    [RDM-5398](https://tools.hmcts.net/jira/browse/RDM-5398).

This PR contains big changes in generic components. Regression test is needed .

### CODE REVIEW

File: src/shared/components/palette/collection/write-collection-field.component.ts [hunk 1]
```diff
@@ -43,8 +43,8 @@ export class WriteCollectionFieldComponent extends AbstractFieldWriteComponent i
       this.profileNotifier.profile.subscribe(_ => this.profile = _);
     }
     this.caseField.value = this.caseField.value || [];
-
     this.formArray = this.registerControl(new FormArray([]));
+    this.formArray.setErrors(null);
```
Comment (mario-paniccia) [inline src/shared/components/palette/collection/write-collection-field.component.ts]: please explain in details why this change

File: src/shared/components/palette/collection/write-collection-field.html [hunk 2]
```diff
@@ -8,7 +8,6 @@ <h2 class="heading-h2">
     <button class="button" type="button" (click)="addItem(true)" [disabled]="isNotAuthorisedToCreate()">Add new</button>
     <h2 class="heading-h2 error-spacing">
       <span *ngIf="caseField.hint_text" class="form-hint">{{caseField.hint_text}}</span>
-      <span *ngIf="formArray.errors" class="error-message">{{formArray.errors | ccdFirstError}}</span>
```
Comment (mario-paniccia) [inline src/shared/components/palette/collection/write-collection-field.html]: please explain in details why this change

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 3]
```diff
@@ -198,9 +220,17 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
...
     if (this.uploadedDocument) {
-      return this.uploadedDocument.get('document_filename').value;
+      return this.uploadedDocument.get(WriteDocumentFieldDefinitions.DOCUMENT_FILENAME).value;
     } else {
       return undefined;
     }
   }
 }
+
+class WriteDocumentFieldDefinitions {
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: do we need a new class here? I think you could just declare those as `const` at the beginning of the file. Check for example: https://github.com/hmcts/ccd-case-ui-toolkit/blob/1241a6146fa1af8483545d11648510aef5d8e03e/src/shared/components/search-filters/search-filters.component.ts#L7

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 4]
```diff
@@ -136,35 +145,48 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
...
     }
   }
-
-  private createDocumentGroup(url?: string, binaryUrl?: string, filename?: string): void {
-    this.uploadedDocument = this.registerControl(new FormGroup({
-      document_url: new FormControl(url || '', Validators.required),
-      document_binary_url: new FormControl(binaryUrl || '', Validators.required),
-      document_filename: new FormControl(filename || '', Validators.required)
-    }));
+  private setComponentToInitialValues() {
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: this method name is a bit misleading. According to the implementation, this should be called: `setComponentToInitialValuesIfMandatory`

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 5]
```diff
@@ -136,35 +145,48 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
       });
     } else {
 
-      this.selectedFile = null;
-      this.uploadedDocument.get('document_url').setValue(null);
-      this.uploadedDocument.get('document_binary_url').setValue(null);
-      this.uploadedDocument.get('document_filename').setValue(null);
-
+      this.setComponentToInitialValues();
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: there is a change of behaviour here. The method `setComponentToInitialValues` executes only if this is a mandatory component. But the code above was executing always. Is this what we want?

File: src/shared/components/palette/document/write-document-field.component.ts (hunk 4 above)
Comment (mario-paniccia) [reply src/shared/components/palette/document/write-document-field.component.ts]: I'd call this `setMandatoryDocumentToEmtpyValues`. That's what the method is doing

File: src/shared/components/palette/collection/write-collection-field.html (hunk 2 above)
Comment (Thor-tech-of-metal) [reply src/shared/components/palette/collection/write-collection-field.html]: DONE change reverted.

File: src/shared/components/palette/document/write-document-field.component.ts (hunk 4 above)
Comment (Thor-tech-of-metal) [reply src/shared/components/palette/document/write-document-field.component.ts]: DONE

File: src/shared/components/palette/document/write-document-field.component.ts (hunk 3 above)
Comment (Thor-tech-of-metal) [reply src/shared/components/palette/document/write-document-field.component.ts]: DONE I created class members  static readonly UPLOAD_ERROR_FILE_REQUIRED

File: src/shared/components/palette/collection/write-collection-field.component.ts (hunk 1 above)
Comment (Thor-tech-of-metal) [reply src/shared/components/palette/collection/write-collection-field.component.ts]: DONE change reverted.

File: src/shared/components/palette/document/write-document-field.component.spec.ts [hunk 6]
```diff
@@ -148,12 +148,6 @@ describe('WriteDocumentFieldComponent', () => {
     expect(ccdReadDocumentElement).toBeTruthy();
   });
 
-  it('should initialise formControl with provided value', () => {
-    expect(FORM_GROUP.controls[FORM_GROUP_ID].value.document_url).toBe(VALUE.document_url);
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.spec.ts]: how come this test has been removed? it seems a good test to me

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 7]
```diff
@@ -118,10 +143,13 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
       documentUpload.append('files', this.selectedFile, this.selectedFile.name);
       documentUpload.append('classification', 'PUBLIC');
       this.documentManagement.uploadFile(documentUpload).subscribe(result => {
-        if (!this.uploadedDocument ) {
-          this.createDocumentGroup();
+        if (!this.uploadedDocument) {
+            this.uploadedDocument = this.registerControl(new FormGroup({
+              document_url: new FormControl(WriteDocumentFieldComponent.DOCUMENT_URL),
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: can u reuse `setDocumentFormValues` method?
Furthermore, why we setting the values to `WriteDocumentFieldComponent.DOCUMENT_URL`, `WriteDocumentFieldComponent.DOCUMENT_BINARY_URL` and `WriteDocumentFieldComponent.DOCUMENT_FILENAME`?

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 8]
```diff
@@ -136,35 +164,30 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
       });
     } else {
 
-      this.selectedFile = null;
-      this.uploadedDocument.get('document_url').setValue(null);
-      this.uploadedDocument.get('document_binary_url').setValue(null);
-      this.uploadedDocument.get('document_filename').setValue(null);
-
+      this.setMandatoryDocumentToEmtpyValues();
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: I think you can remove `this.setMandatoryDocumentToEmtpyValues();` and the method `setMandatoryDocumentToEmtpyValues()`  and reuse the ones we introduced. Change to:
```if (this.isAMandatoryComponent()) {
       this.selectedFile = null;
        this.setDocumentFormValuesWithValidator(null, null, null);
        this.displayFileErrors();
      }```

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 9]
```diff
@@ -55,16 +60,36 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
...
+      }
+    } else {
+      if (this.isAMandatoryComponent()) {
+        this.setDocumentFormValuesWithValidator(null, null, null);
+        this.selectedFile = null;
+      }
     }
   }
 
+  private setDocumentFormValuesWithValidator(url: string, binaryUrl: string, filename: string) {
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: can you move these 2 private methods down after the public methods?

File: src/shared/components/palette/document/write-document-field.component.ts [hunk 10]
```diff
@@ -198,9 +139,75 @@ export class WriteDocumentFieldComponent extends AbstractFieldWriteComponent imp
...
+  private isAMandatoryComponent() {
+    return this.caseField.display_context && this.caseField.display_context === Constants.MANDATORY;
+  }
+
+  private displayFileErrors () {
+    this.valid = false;
+    this.uploadError = WriteDocumentFieldComponent.UPLOAD_ERROR_FILE_REQUIRED;
+  }
+
+  private  validateFormUploadedDocument():  boolean {
```
Comment (mario-paniccia) [inline src/shared/components/palette/document/write-document-field.component.ts]: extra space here to be removed after `private`

**Comment by hemantt** [comment]
Closing the stale PR
